File: NetFramework/DashboardLayout/VB/DashboardLayoutDemo/DemoControls/BaseForm.vb
Imports System
Imports System.ComponentModel
Imports C1.Win.C1Ribbon
Imports C1.Win.C1Themes
Imports C1.Win.Layout

Namespace DashboardDemo.Controls
    Public Partial Class BaseForm
        Inherits C1RibbonForm
        Implements IDemo

        Private _layoutType As LayoutType

        Public Sub New()
            InitializeComponent()
        End Sub

        Public Property RibbonVisualStyle As VisualStyle
            Get
                Return c1Ribbon1.VisualStyle
            End Get
            Set(ByVal value As VisualStyle)
                c1Ribbon1.VisualStyle = value
            End Set
        End Property

        Public Property ToolIconAppearance As ToolIconAppearance

        Public Property LayoutType As LayoutType
            Get
                Return _layoutType
            End Get
            Set(ByVal value As LayoutType)

                If _layoutType <> value Then
                    _layoutType = value

                    Select Case _layoutType
                        Case LayoutType.Flow
                            btnFlow.Pressed = True
                        Case LayoutType.Grid
                            btnGrid.Pressed = True
                        Case LayoutType.Split
                            btnSplit.Pressed = True
                    End Select
                End If
            End Set
        End Property

        <Category("Sample")>
        <Description("Occurs when ItemIndex of Theme ComboBox is changed.")>
        Public Event DemoThemeChanged As EventHandler(Of ThemeEventArgs)

        Protected Sub OnDemoThemeChanged(ByVal sender As Object, ByVal themeName As String)
            RaiseEvent DemoThemeChanged(sender, New ThemeEventArgs(themeName))
        End Sub

        <Category("Sample")>
        <Description("Occurs when ItemIndex of ToolIcon ComboBox is changed.")>
        Public Event ToolIconAppearanceChanged As EventHandler(Of ToolIconAppearanceEventArgs)

        Protected Sub OnToolIconAppearanceChanged(ByVal sender As Object, ByVal tia As ToolIconAppearance)
            RaiseEvent ToolIconAppearanceChanged(sender, New ToolIconAppearanceEventArgs(tia))
        End Sub

        <Category("Sample")>
        <Description("Occurs when one of the apply layout menu button is clicked.")>
        Public Event ApplyLayoutClicked As EventHandler(Of FilePathEventArgs)

        Protected Sub OnApplyLayoutClicked(ByVal sender As Object, ByVal e As EventArgs)
            Dim btn As RibbonItem = TryCast(sender, RibbonItem)
            If btn IsNot Nothing Then RaiseEvent ApplyLayoutClicked(sender, New FilePathEventArgs(SampleHelper.Instance.GetLayoutPath(TryCast(btn.Tag, String))))
        End Sub

        <Category("Sample")>
        <Description("Occurs when save layout menu button is clicked.")>
        Public Event SaveLayoutClicked As EventHandler(Of FilePathEventArgs)

        Protected Sub OnSaveLayoutClicked(ByVal sender As Object, ByVal path As String)
            RaiseEvent SaveLayoutClicked(sender, New FilePathEventArgs(path))
        End Sub

        <Category("Sample")>
        <Description("Occurs when reset layout menu button is clicked.")>
        Public Event ResetLayoutClicked As EventHandler(Of FilePathEventArgs)

        Protected Sub OnResetLayoutClicked(ByVal sender As Object, ByVal path As String)
            RaiseEvent ResetLayoutClicked(sender, New FilePathEventArgs(path))
        End Sub

        <Category("Sample")>
        <Description("Occurs when one of the LyoutType toggle button is pressed.")>
        Public Event LayoutTypeChanged As EventHandler(Of LayoutTypeEventArgs)

        Protected Sub OnLayoutTypeChanged(ByVal sender As Object, ByVal type As LayoutType)
            RaiseEvent LayoutTypeChanged(sender, New LayoutTypeEventArgs(type))
        End Sub

        Private Sub LoadThemes()
            cmbThemes.Items.Clear()
            Dim themes = C1ThemeController.GetThemes()
            cmbThemes.Items.Add("None")

            For Each theme In themes
                If Not theme.Contains("Office2010") AndAlso Not theme.Contains("Office2007") Then cmbThemes.Items.Add(theme)
            Next

            cmbThemes.SelectedIndex = cmbThemes.Items.IndexOf("Office2016Colorful")
        End Sub

        Private Sub LoadIconAppearance()
            cmbToolicon.Items.Clear()
            Dim positions = [Enum].GetValues(GetType(ToolIconAppearance))

            For i = 0 To positions.Length - 1
                Dim value = positions.GetValue(i).ToString()
                cmbToolicon.Items.Add(value)
                If value.Equals(ToolIconAppearance.ToString()) Then cmbToolicon.SelectedIndex = i
            Next
        End Sub

        Private Sub UpdateLayouts()
            For Each item As RibbonButton In menuApply.Items
                RemoveHandler item.Click, AddressOf OnApplyLayoutClicked
            Next

            menuApply.Items.Clear()
            Dim layouts As String() = SampleHelper.Instance.GetLayouts()

            If layouts IsNot Nothing Then

                If layouts.Length = 0 Then
                    menuApply.Enabled = False
                Else
                    menuApply.Enabled = True

                    For Each item In layouts
                        Dim btn = New RibbonButton()
                        btn.Text = item.Substring(0, item.LastIndexOf("."c))
                        btn.Tag = item
                        AddHandler btn.Click, AddressOf OnApplyLayoutClicked
                        menuApply.Items.Add(btn)
                    Next
                End If
            Else
                menuApply.Enabled = False
            End If
        End Sub

        Private Sub menuSave_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim filePath As String

            If SampleHelper.Instance.SaveLayoutDialog(filePath) Then
                OnSaveLayoutClicked(sender, filePath)
                UpdateLayouts()
            End If
        End Sub

        Private Sub menuReset_Click(ByVal sender As Object, ByVal e As EventArgs)
            OnResetLayoutClicked(sender, SampleHelper.Instance.DefaultLayoutPath)
        End Sub

        Private Sub menuManage_Click(ByVal sender As Object, ByVal e As EventArgs)
            SampleHelper.Instance.ManageLayoutDialog()
            UpdateLayouts()
        End Sub

        Private Sub btnFlow_PressedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If btnFlow.Pressed Then OnLayoutTypeChanged(sender, LayoutType.Flow)
        End Sub

        Private Sub btnGrid_PressedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If btnGrid.Pressed Then OnLayoutTypeChanged(sender, LayoutType.Grid)
        End Sub

        Private Sub btnSplit_PressedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If btnSplit.Pressed Then OnLayoutTypeChanged(sender, LayoutType.Split)
        End Sub

        Private Sub cmbToolicon_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            OnToolIconAppearanceChanged(sender, CType([Enum].Parse(GetType(ToolIconAppearance), cmbToolicon.SelectedItem.Text), ToolIconAppearance))
        End Sub

        Private Sub cmbThemes_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            OnDemoThemeChanged(sender, cmbThemes.SelectedItem.Text)
        End Sub

        Public Sub Init() Implements IDemo.Init
            LoadThemes()
            LoadIconAppearance()
            UpdateLayouts()
        End Sub
    End Class
End Namespace

File: NetFramework/DashboardLayout/CS/DashboardLayoutDemo/DemoControls/BaseForm.cs
using System;
using System.ComponentModel;
using C1.Win.C1Ribbon;
using C1.Win.C1Themes;
using C1.Win.Layout;

namespace DashboardDemo.Controls
{
    public partial class BaseForm : C1RibbonForm, IDemo
    {
        #region Fields
        private LayoutType _layoutType;
        #endregion Fields

        #region Ctor
        public BaseForm()
        {
            InitializeComponent();
        }
        #endregion Ctor

        #region Properties
        public VisualStyle RibbonVisualStyle
        {
            get { return c1Ribbon1.VisualStyle; }
            set { c1Ribbon1.VisualStyle = value; }
        }

        public ToolIconAppearance ToolIconAppearance { get; set; }

        public LayoutType LayoutType
        {
            get { return _layoutType; }
            set
            {
                if (_layoutType != value)
                {
                    _layoutType = value;
                    switch (_layoutType)
                    {
                        case LayoutType.Flow:
                            btnFlow.Pressed = true;
                            break;
                        case LayoutType.Grid:
                            btnGrid.Pressed = true;
                            break;
                        case LayoutType.Split:
                            btnSplit.Pressed = true;
                            break;
                    }
                }
            }
        }
        #endregion Properties               

        #region Events
        /// <summary>
        /// Occurs when ItemIndex of Theme ComboBox is changed.
        /// </summary>
        [Category("Sample")]
        [Description("Occurs when ItemIndex of Theme ComboBox is changed.")]
        public event EventHandler<ThemeEventArgs> DemoThemeChanged;
        protected void OnDemoThemeChanged(object sender, string themeName)
        {
            DemoThemeChanged?.Invoke(sender, new ThemeEventArgs(themeName));
        }
        /// <summary>
        /// Occurs when ItemIndex of ToolIcon ComboBox is changed.
        /// </summary>
        [Category("Sample")]
        [Description("Occurs when ItemIndex of ToolIcon ComboBox is changed.")]
        public event EventHandler<ToolIconAppearanceEventArgs> ToolIconAppearanceChanged;
        protected void OnToolIconAppearanceChanged(object sender, ToolIconAppearance tia)
        {
            ToolIconAppearanceChanged?.Invoke(sender, new ToolIconAppearanceEventArgs(tia));
        }
        /// <summary>
        /// Occurs when one of the apply layout menu button is clicked.
        /// </summary>
        [Category("Sample")]
        [Description("Occurs when one of the apply layout menu button is clicked.")]
        public event EventHandler<FilePathEventArgs> ApplyLayoutClicked;
        protected void OnApplyLayoutClicked(object sender, EventArgs e)
        {
            RibbonItem btn = sender as RibbonItem;
            if (btn != null)
                ApplyLayoutClicked?.Invoke(sender, new FilePathEventArgs(SampleHelper.Instance.GetLayoutPath(btn.Tag as string)));
        }
        /// <summary>
        /// Occurs when save layout menu button is clicked.
        /// </summary>
        [Category("Sample")]
        [Description("Occurs when save layout menu button is clicked.")]
        public event EventHandler<FilePathEventArgs> SaveLayoutClicked;
        protected void OnSaveLayoutClicked(object sender, string path)
        {
            SaveLayoutClicked?.Invoke(sender, new FilePathEventArgs(path));
        }
        /// <summary>
        /// Occurs when reset layout menu button is clicked.
        /// </summary
        [Category("Sample")]
        [Description("Occurs when reset layout menu button is clicked.")]
        public event EventHandler<FilePathEventArgs> ResetLayoutClicked;
        protected void OnResetLayoutClicked(object sender, string path)
        {
            ResetLayoutClicked?.Invoke(sender, new FilePathEventArgs(path));
        }
        /// <summary>
        /// Occurs when one of the LyoutType toggle button is pressed.
        /// </summary
        [Category("Sample")]
        [Description("Occurs when one of the LyoutType toggle button is pressed.")]
        public event EventHandler<LayoutTypeEventArgs> LayoutTypeChanged;
        protected void OnLayoutTypeChanged(object sender, LayoutType type)
        {
            LayoutTypeChanged?.Invoke(sender, new LayoutTypeEventArgs(type));
        }
        #endregion        

        #region Private
        private void LoadThemes()
        {
            cmbThemes.Items.Clear();

            var themes = C1ThemeController.GetThemes();
            cmbThemes.Items.Add("None");
            foreach (var theme in themes)
                if(!theme.Contains("Office2010") && !theme.Contains("Office2007"))
                    cmbThemes.Items.Add(theme);
            cmbThemes.SelectedIndex = cmbThemes.Items.IndexOf("Office2016Colorful");
        }

        private void LoadIconAppearance()
        {
            cmbToolicon.Items.Clear();

            var positions = Enum.GetValues(typeof(ToolIconAppearance));
            for (var i = 0; i < positions.Length; i++)
            {
                var value = positions.GetValue(i).ToString();
                cmbToolicon.Items.Add(value);
                if (value.Equals(ToolIconAppearance.ToString()))
                    cmbToolicon.SelectedIndex = i;
            }
        }

        private void UpdateLayouts()
        {
            foreach (RibbonButton item in menuApply.Items)
                item.Click -= OnApplyLayoutClicked;
            menuApply.Items.Clear();

            string[] layouts = SampleHelper.Instance.GetLayouts();
            if (layouts != null)
            {
                if (layouts.Length == 0)
                {
                    menuApply.Enabled = false;
                }
                else
                {
                    menuApply.Enabled = true;
                    foreach (var item in layouts)
                    {
                        var btn = new RibbonButton();
                        btn.Text = item.Substring(0, item.LastIndexOf('.'));
                        btn.Tag = item;
                        btn.Click += OnApplyLayoutClicked;
                        menuApply.Items.Add(btn);
                    }
                }
            }
            else
                menuApply.Enabled = false;
        }
        #endregion Private

        #region EventHandlers
        private void menuSave_Click(object sender, EventArgs e)
        {
            string filePath;
            if (SampleHelper.Instance.SaveLayoutDialog(out filePath))
            {
                OnSaveLayoutClicked(sender, filePath);
                UpdateLayouts();
            }
        }

        private void menuReset_Click(object sender, EventArgs e)
        {
            OnResetLayoutClicked(sender, SampleHelper.Instance.DefaultLayoutPath);            
        }

        private void menuManage_Click(object sender, EventArgs e)
        {
            SampleHelper.Instance.ManageLayoutDialog();
            UpdateLayouts();
        }

        private void btnFlow_PressedChanged(object sender, EventArgs e)
        {
            if (btnFlow.Pressed)
                OnLayoutTypeChanged(sender, LayoutType.Flow);
        }

        private void btnGrid_PressedChanged(object sender, EventArgs e)
        {
            if (btnGrid.Pressed)
                OnLayoutTypeChanged(sender, LayoutType.Grid);
        }

        private void btnSplit_PressedChanged(object sender, EventArgs e)
        {
            if (btnSplit.Pressed)
                OnLayoutTypeChanged(sender, LayoutType.Split);
        }

        private void cmbToolicon_SelectedIndexChanged(object sender, EventArgs e)
        {
            OnToolIconAppearanceChanged(sender, (ToolIconAppearance)Enum.Parse(typeof(ToolIconAppearance), cmbToolicon.SelectedItem.Text));
        }

        private void cmbThemes_SelectedIndexChanged(object sender, EventArgs e)
        {
            OnDemoThemeChanged(sender, cmbThemes.SelectedItem.Text);
        }
        #endregion EventHandlers

        #region IDemo
        public void Init()
        {
            LoadThemes();
            LoadIconAppearance();
            UpdateLayouts();
        }
        #endregion IDemo
    }
}
